File: src/cleansia_android/core/src/main/java/cz/cleansia/core/freshness/Staleness.kt
package cz.cleansia.core.freshness

import java.util.concurrent.atomic.AtomicLong

/**
 * Per-cache freshness watermark. Repositories hold one per logical cache and mark it after a successful
 * fetch; ViewModels check it before triggering a background refresh on screen entry.
 *
 * **User-initiated pulls bypass it entirely** — the user's intent is the source of truth, not cache age.
 * Orthogonal to the session-scoped wipe: a repo may hold both and should reset this from its existing
 * wipe hook. -> /mobile-app/patterns#session-wipe
 */
class Staleness {
    private val lastFetchedAtMillis = AtomicLong(NEVER)

    /** Epoch millis of the last successful fetch, or `null` if never fetched. */
    val lastFetchedAt: Long?
        get() = lastFetchedAtMillis.get().takeIf { it != NEVER }

    /** Stamp the watermark with the current time. Call after a successful fetch. */
    fun markFresh() {
        lastFetchedAtMillis.set(System.currentTimeMillis())
    }

    /**
     * `true` if the cache has never been fetched, or if it was last fetched
     * more than [maxAgeMs] ago. Default is [DEFAULT_MAX_AGE_MS].
     */
    fun isStale(maxAgeMs: Long = DEFAULT_MAX_AGE_MS): Boolean {
        val stamp = lastFetchedAtMillis.get()
        if (stamp == NEVER) return true
        return (System.currentTimeMillis() - stamp) > maxAgeMs
    }

    /** Forget the watermark. Call from `SessionScopedCache.clear()` on logout. */
    fun reset() {
        lastFetchedAtMillis.set(NEVER)
    }

    companion object {
        const val DEFAULT_MAX_AGE_MS: Long = 30_000L
        private const val NEVER = 0L
    }
}

File: src/cleansia_ios/CleansiaCore/Sources/CleansiaCore/State/Staleness.swift
import Foundation

/// Freshness watermark for one logical cache: stamp after a successful fetch, ask before spending a
/// round-trip.
///
/// **User-initiated pulls bypass it entirely** — the user's intent outranks cache age. **A "loaded" flag
/// cannot double as a freshness signal**: it is a one-way first-paint latch, so anything that changes
/// server-side after the first fetch never reaches the screen.
/// → /mobile-app/patterns#session-wipe
@MainActor
public final class Staleness {
    private let window: TimeInterval
    private let now: () -> Date
    private var mark: Date?

    /// `nonisolated` so it can be a default argument. Repositories declare
    /// `staleness: Staleness = Staleness()`, and a default argument is evaluated in the
    /// CALLER's context, which is nonisolated — an isolated init there is a compile error
    /// at every call site rather than at the declaration, which is why it is easy to miss.
    /// Safe because the initializer only assigns stored properties; every subsequent access
    /// is still `MainActor`-isolated.
    public nonisolated init(window: TimeInterval = 30, now: @escaping () -> Date = Date.init) {
        self.window = window
        self.now = now
    }

    public var isStale: Bool {
        guard let mark else { return true }
        return now().timeIntervalSince(mark) >= window
    }

    /// Stamp the watermark. Call only after a fetch that actually landed — a
    /// failed fetch left the cache untouched, so the next entry must retry.
    public func markFresh() {
        mark = now()
    }

    /// Forget the watermark: the next check reads stale. Sign-out, and any
    /// mutation the owner knows invalidated the cache.
    public func invalidate() {
        mark = nil
    }
}
